server/services/v1/user_metadata_provider.go

```diff
@@ -0,0 +1,148 @@
+		if err = tx.Rollback(ctx); err != nil {
+			log.Debug().Msg("Failed to rollback transaction.")
+		}
+		return nil, api.Errorf(api.Code_INTERNAL, "Failed to read user metadata.")
+	}
+
+	return &api.GetUserMetadataResponse{
+		MetadataKey: req.GetMetadataKey(),
+		UserId:      currentSub,
+		NamespaceId: namespaceId,
+		Value: &api.UserMetadataValue{
+			Value: val,
+		},
+	}, nil
+}
+
+func (a *DefaultUserMetadataProvider) InsertUserMetadata(ctx context.Context, req *api.InsertUserMetadataRequest) (*api.InsertUserMetadataResponse, error) {
+	namespace, err := request.GetNamespace(ctx)
+	if err != nil {
+		return nil, api.Errorf(api.Code_INTERNAL, "Failed to insert user metadata. reason: failed to read user namespace.")
+	}
+
+	currentSub, err := getCurrentSub(ctx)
+	if err != nil {
+		return nil, api.Errorf(api.Code_INTERNAL, "Failed to insert user metadata. reason: failed to read user.")
+	}
+
+	namespaceId, err := a.tenantMgr.GetNamespaceId(namespace)
+	if err != nil {
+		return nil, api.Errorf(api.Code_INTERNAL, "Failed to insert user metadata. reason: failed to read namespace id.")
+	}
+
+	tx, err := a.txMgr.StartTx(ctx)
+	if err != nil {
+		return nil, api.Errorf(api.Code_INTERNAL, "Failed to insert user metadata. reason: failed to create internal transaction.")
+	}
+
+	err = a.userStore.InsertUserMetadata(ctx, tx, namespaceId, metadata.User, currentSub, req.GetMetadataKey(), req.GetValue().GetValue())
+	if err != nil {
+		if err = tx.Rollback(ctx); err != nil {
+			log.Debug().Msg("Failed to rollback transaction.")
+		}
+		return nil, api.Errorf(api.Code_INTERNAL, "Failed to insert user metadata.")
+	}
+	if err = tx.Commit(ctx); err != nil {
+		log.Debug().Msg("Failed to commit transaction.")
+		return nil, api.Errorf(api.Code_INTERNAL, "Failed to insert user metadata. reason: transaction was not committed.")
+	}
+	return &api.InsertUserMetadataResponse{
+		MetadataKey: req.GetMetadataKey(),
+		UserId:      currentSub,
+		NamespaceId: namespaceId,
+		Value: &api.UserMetadataValue{
+			Value: req.GetValue().GetValue(),
+		},
+	}, nil
+}
+
+func (a *DefaultUserMetadataProvider) UpdateUserMetadata(ctx context.Context, req *api.UpdateUserMetadataRequest) (*api.UpdateUserMetadataResponse, error) {
+	namespace, err := request.GetNamespace(ctx)
+	if err != nil {
+		return nil, api.Errorf(api.Code_INTERNAL, "Failed to update user metadata. reason: failed to read user namespace.")
+	}
+
+	currentSub, err := getCurrentSub(ctx)
+	if err != nil {
+		return nil, api.Errorf(api.Code_INTERNAL, "Failed to update user metadata. reason: failed to read user.")
+	}
+
+	namespaceId, err := a.tenantMgr.GetNamespaceId(namespace)
+	if err != nil {
+		return nil, api.Errorf(api.Code_INTERNAL, "Failed to update user metadata. reason: failed to read namespace id.")
+	}
+
+	tx, err := a.txMgr.StartTx(ctx)
+	if err != nil {
+		return nil, api.Errorf(api.Code_INTERNAL, "Failed to update user metadata. reason: failed to create internal transaction.")
+	}
```

Comment:
The code from the beginning of the function up to this point is identical in three above functions can we dedup?

Reply:
refactored.